lms/djangoapps/course_api/blocks/views.py

```diff
@@ -303,3 +303,93 @@ def list(self, request, hide_access_denials=False):  # pylint: disable=arguments
             raise ValidationError(u"'{}' is not a valid course key.".format(six.text_type(course_key_string)))
         return super(BlocksInCourseView, self).list(request, course_usage_key, hide_access_denials=hide_access_denials)
+
+
+@view_auth_classes(is_authenticated=False)
+class BlocksWithCompletionView(BlocksInCourseView):
+    """
+    **Use Case**
+
+        Returns the blocks in the course according to the requesting user's
+        access level.
+
+    **Example requests**:
+
+        GET /api/courses/v1/blocks/?course_id=<course_id>
+        GET /api/courses/v1/blocks/?course_id=<course_id>
+            &username=anjali
+            &depth=all
+            &requested_fields=graded,format,student_view_multi_device,lti_url
+            &block_counts=video
+            &student_view_data=video
+            &block_types_filter=problem,html
+
+    **Parameters**:
+
+        This view redirects to /api/courses/v2/blocks/<root_usage_key>/ for the
+        root usage key of the course specified by course_id.  The view accepts
+        all parameters accepted by :class:`BlocksInCourseView`
+
+    **Response Values**
+
+        Responses are identical to those returned by :class:`BlocksInCourseView` when
+        passed the root_usage_key of the requested course but there is one exception explained below.
+
+        It adds completion from root of the course to every child block and make it boolean i.e
+        completion will be either 1 or 0
+
+    """
+
+    def list(self, request, hide_access_denials=False):
+        """
+        Retrieves the usage_key for the requested course, and then returns the
+        same information that would be returned by BlocksView.list, called with
+        that usage key
+
+        Arguments:
+            request - Django request object
+        """
+
+        response = super().list(request, hide_access_denials=hide_access_denials)
+
+        course_blocks = {}
+        if request.query_params.get('return_type') == 'list':
+            for course_block in response.data:
+                course_blocks[course_block['id']] = course_block
+
+                if course_block.get('type') == 'course':
+                    root = course_block['id']
+        else:
+            root = response.data['root']
+            course_blocks = response.data['blocks']
+
+        recurse_mark_complete(root, course_blocks)
+
+        return response
+
+
+def recurse_mark_complete(block_id, blocks):
+    """
+    Helper function to walk course tree dict,
+    marking completion as 1 or 0
+
+    If all blocks are complete, mark parent block complete
+
+    :param blocks: dict of all blocks
+    :param block_id: root or child block id
+
+    :return:
+        block: course_outline_root_block block object or child block
+    """
+    block = blocks.get(block_id, {})
+    if block.get('completion') == 1:
+        return
+
+    child_blocks = block.get('children', block.get('descendents'))
+    if child_blocks:
```

Comment:
if there are no children, then completion will not be marked. is that as intended?

Reply:
yes, this is intended. Leaf blocks or units completion is being marked by patch call to completion service.

Reply:
makes sense. please note that in a comment.

Reply:
@jawad-khan could you note this in a comment?

Reply:
Done.
https://github.com/edx/edx-platform/pull/26079/files#diff-b22862b300f8aa061c55c284749f63b952becb99b16790afa8fcb60fd1d56ce2R350